tutorials/common/common_utils.h

```diff
@@ -117,3 +117,157 @@ void mfxGetTime(mfxTime* timestamp);
+{
+    mfxU16             m_numSurfaces    = 0;
+    mfxFrameSurface1** m_pmfxSurfaces   = nullptr;
+
+    appResourcesVidMem () {}
+
+
+    virtual ~appResourcesVidMem()
+    {
+
+        for (int i = 0; i < m_numSurfaces; i++)
+            delete m_pmfxSurfaces[i];
+
+        MSDK_SAFE_DELETE_ARRAY(m_pmfxSurfaces);
+
+    }
+
+    // forbidden
+    appResourcesVidMem (const appResourcesVidMem& res)            = delete;
+    appResourcesVidMem &operator= (const appResourcesVidMem& res) = delete;
+};
+
+struct appResourcesSysMem : appResourcesVidMem
+{
+    mfxU8*             m_surfaceBuffers = nullptr;
+
+    appResourcesSysMem () {}
+
+    virtual ~appResourcesSysMem()
+    {
+        MSDK_SAFE_DELETE_ARRAY(m_surfaceBuffers);
+    }
+
+    // forbidden
+    appResourcesSysMem (const appResourcesSysMem& res)            = delete;
+    appResourcesSysMem &operator= (const appResourcesSysMem& res) = delete;
+};
+
+
+
+struct appResourcesVMemAsync : public appResourcesSysMem, public appResourcesAsync
+{
+    appResourcesVMemAsync() {}
+    virtual ~appResourcesVMemAsync() { }
+    // forbidden
+    appResourcesVMemAsync (const appResourcesVMemAsync& res)            = delete;
+    appResourcesVMemAsync &operator= (const appResourcesVMemAsync& res) = delete;
+};
+
+struct appResourcesVpp : public appResources
+{
+
+    mfxU8* m_surfaceBuffersIn  = nullptr;
+    mfxU8* m_surfaceBuffersOut = nullptr;
+
+    mfxFrameSurface1** m_pVPPSurfacesIn  = nullptr;
+    mfxFrameSurface1** m_pVPPSurfacesOut = nullptr;
+
+    mfxU16 m_nVPPSurfNumIn  = 0;
+    mfxU16 m_nVPPSurfNumOut = 0;
+
+    appResourcesVpp () {}
+
+    virtual ~appResourcesVpp ()
+    {
+        for (int i = 0; i < m_nVPPSurfNumIn; i++)
+            delete m_pVPPSurfacesIn[i];
+
+        MSDK_SAFE_DELETE_ARRAY(m_pVPPSurfacesIn);
+
+        for (int i = 0; i < m_nVPPSurfNumOut; i++)
+            delete m_pVPPSurfacesOut[i];
+
+        MSDK_SAFE_DELETE_ARRAY(m_pVPPSurfacesOut);
+
+        MSDK_SAFE_DELETE_ARRAY(m_surfaceBuffersIn);
+        MSDK_SAFE_DELETE_ARRAY(m_surfaceBuffersOut);
+    }
+
+    // forbidden
+    appResourcesVpp (const appResourcesVpp& res)            = delete;
+    appResourcesVpp &operator= (const appResourcesVpp& res) = delete;
+};
+
+struct appResourcesTranscode : public appResourcesSysMem
+{
+    mfxBitstream m_mfxEncBS;
+
+    appResourcesTranscode() {}
+
+    virtual ~appResourcesTranscode()
+    {
+        MSDK_SAFE_DELETE_ARRAY(m_mfxEncBS.Data);
+    }
+
+    // forbidden
+    appResourcesTranscode (const appResourcesTranscode& res)            = delete;
+    appResourcesTranscode &operator= (const appResourcesTranscode& res) = delete;
+};
+
+struct appResourcesVppAsync : public appResourcesVpp, public appResourcesAsync
+{
+    appResourcesVppAsync() {}
+    ~appResourcesVppAsync() {}
+
+    // forbidden
+    appResourcesVppAsync (const appResourcesVppAsync& res)            = delete;
+    appResourcesVppAsync &operator= (const appResourcesVppAsync& res) = delete;
+};
```

Comment:
Anton, here you're enabling custom classes which in fact are RAII idiom. I suggest avoiding doing this.  Introduction **unnecessary** (please see below) entities, when you can use something well-known and common, is usually a bad practice. Instead of these set of classes you can use:
* std vectors for system memory
* unique_ptr with user-specified cleanup functions for files

Reply:
fully agree

Reply:
Changed.